File: Java/ToyLanguageGUI_exam_final/src/repository/Repository.java
package repository;

import domain.*;

import java.io.*;
import java.util.ArrayList;
import java.util.List;

import domain.statements.Statement;
import exceptions.*;


public class Repository implements IRepository {
    private List<PrgState> a;
    private String logFile;

    public Repository(String fileName){
        logFile = fileName;
        a = new ArrayList<>();
    }

    public void addPrgState(PrgState p){
        a.add(p);
    }


    public void logProgStateExec(PrgState p){
        try(PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter(logFile,true)));
        ){

            pw.println("\n ID:"+p.getId()+"\n");
            pw.println("------------------");
            pw.println("ExeStack: ");
            PrgState current = p;
            for(Statement s:current.getStack().getAll())
                pw.println(s);
            pw.println("\n");
            pw.println("SymTable: ");
            for(String s : current.getSymbols().getAll()) {
                    int val = current.getSymbols().get(s);
                    pw.println(s + "->" + val);
            }
            pw.println("\n");
            pw.println("Out: ");
            for(Integer i : current.getMessages().getAll())
                pw.println(i);
            pw.println("\n");
            pw.println("FileTable: ");
            for(Integer i : current.getFileTable().getAll()) {
                FileData val = current.getFileTable().get(i);
                pw.println(i + "->" +val);
            }
            pw.println("\n");
            pw.println("Heap: ");
            for(Integer k : current.getHeap().getAll()){
                int val = current.getHeap().get(k);
                pw.println(k + "->" + val);
            }
        }
        catch(IOException ioe){
            throw new InterpretorException(ioe.getMessage());
        }
    }

    public List<PrgState> getPrgList(){
        return a;
    }

    public void setPrgList(List<PrgState> l){
        a = l;
    }
}

File: ToyInterpretor/ToyInterpretor/Repository/Repository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ToyInterpretor.Domain;
using ToyInterpretor.Exceptions;
using System.IO;
using ToyInterpretor.Domain.Statements;

namespace ToyInterpretor.Repository
{
    class Repo : IRepository
    {
        private List<PrgState> a;
        private String logFile;

        public Repo(String fileName)
        {
            logFile = fileName;
            a = new List<PrgState>();
        }

        public void addPrgState(PrgState p)
        {
            a.Add(p);
        }

        public PrgState getCurrentPrgState()
        {
            if (a.Count <= 0)
                throw new RepoException();
            return a[0];
        }

        public void logProgStateExec()
        {
            try 
             {
                TextWriter tw = new StreamWriter(new FileStream(logFile, FileMode.Append));
                tw.WriteLine("------------------");
                tw.WriteLine("ExeStack: ");
                PrgState current = getCurrentPrgState();
                foreach (Statement s in current.getStack().getAll()) {
                    tw.WriteLine(s);
                }
                tw.WriteLine("\n");
                tw.WriteLine("SymTable: ");
                foreach (String s in current.getSymbols().getAll())
                {
                    int val = current.getSymbols().get(s);
                    tw.WriteLine(s + "->" + val);
                }
                tw.WriteLine("\n");
                tw.WriteLine("Out: ");
                foreach (int i in current.getMessages().getAll())
                {
                    tw.WriteLine(i);
                }
                tw.WriteLine("\n");
                tw.WriteLine("FileTable: ");
                foreach (int i in current.getFileTable().getAll())
                {
                    FileData val = current.getFileTable().get(i);
                    tw.WriteLine(i + "->" + val);
                }
                tw.WriteLine("\n");
               
                tw.Close();
            }
            catch (IOException ioe)
            {
                throw new Exception(ioe.Message);
            }
        }
    }
}
